source/base/fileinputoutput.cpp

```diff
@@ -301,7 +302,7 @@ void OStream::printf(const char *format, ...)
     char buffer[1024];
 
     va_start(marker, format);
-    vsnprintf(buffer, 1023, format, marker);
+    std::vsnprintf(buffer, sizeof(buffer), format, marker);
```

Comment:
Previously, (almost) all invocations of `vsnprintf` and `snprintf` passed the buffer capacity _minus 1_, i.e. _excluding_ room for the terminating NUL character, despite the standard C99 functions expecting the buffer capacity to _include_ room for the NUL. This was done because we were de-facto using `_vsnprintf` and `_snprintf` on Windows machines, which did not follow the standard, and this pattern of invocation was considered safe either way (although on Windows machines it relied on the buffer being zero-initialized).
As we're now using the C++11 functions `std::vsnprintf` and `std::snprintf`, which follow the C99 convention even on Windows machines, we can now safely pass the full size of the buffer.

Also, instead of a literal number we're now using `sizeof(buffer)` wherever `buffer` is a fixed-size char array.

Reply:
Danger, vsnprintf does NOT put a terminating null when the result need to be truncated due to length constraint. (hence the 1023 used previously for buffer[1024]).

Both codes (before & after) are bogus, because it is missing a buffer[1023]='\0'; (before va_start, if vsnprintf() is called with sizeof(buffer)-1 ) to handle the situation

Reply:
What you're writing was (and still is) true for Microsoft's implementation of `_vsnprintf` (note the leading underscore). However, with the introduction of C++11 Microsoft has also introduced a `vsnprintf` function, which (at least according to their docs) _does_ always NUL-terminate the string (as mandated by the C99 standard, which C++11 in turn refers to).

Reply:
Yes, fine since C99/Posix clarification.
I'm too old, and used to be burnt by vsnprintf in old libraries.